### PR TITLE
Add UncaughtExceptionHandler

### PR DESCRIPTION
**Pull request by HubertJaworski**

### CODE REVIEW

File: neptune/new/internal/utils/uncaught_exception_handler.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    def register(self, uid: uuid.UUID, handler: Callable[[List[str]], None]):
+        self._handlers[uid] = handler
+
+    def unregister(self, uid: uuid.UUID):
+        if uid in self._handlers:
+            del self._handlers[uid]
+
+
+instance = UncaughtExceptionHandler()
```
Comment (aniezurawski) [inline neptune/new/internal/utils/uncaught_exception_handler.py:61]: Is this object thread-safe?

File: neptune/new/internal/utils/traceback_job.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        self._started = False
+        self._path = path
+
+    def start(self, run: 'Run'):
+        if not self._started:
+            path = self._path
+
+            def log_traceback(stacktrace_lines: List[str]):
+                run[path].log(stacktrace_lines)
+                run["sys/failed"] = Boolean(True)
```
Comment (aniezurawski) [inline neptune/new/internal/utils/traceback_job.py:45]: Use constant.

File: neptune/new/internal/utils/traceback_job.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+            traceback_handler.register(self._uuid, log_traceback)
+        self._started = True
+
+    def stop(self):
+        traceback_handler.unregister(self._uuid)
+
+    def join(self, seconds: Optional[float] = None):
+        self.stop()
```
Comment (aniezurawski) [inline neptune/new/internal/utils/traceback_job.py:54]: It should be empty. stop is called explicitly before join.

**Review by aniezurawski — CHANGES_REQUESTED** [review]

File: neptune/new/internal/utils/uncaught_exception_handler.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        self._handlers = dict()
+
+    def activate(self):
+
+        def exception_handler(exc_type, exc_val, exc_tb):
+            traceback_lines = traceback.format_tb(exc_tb) + [repr(exc_val)]
+            for _, handler in self._handlers.items():
+                handler(traceback_lines)
+
+            sys.__excepthook__(exc_type, exc_val, exc_tb)
```
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply neptune/new/internal/utils/uncaught_exception_handler.py:42]: Copied from existing code. There was a call to `sys.__excepthook__`

File: neptune/new/internal/init_impl.py [hunk 5]
```diff
@@ -214,7 +215,7 @@ def init(project: Optional[str] = None,
     if tags is not None:
         _run[attr_consts.SYSTEM_TAGS_ATTRIBUTE_PATH].add(tags)
     if run is None:
-        _run[attr_consts.SYSTEM_FAILED_ATTRIBUTE_PATH] = False
+        _run[attr_consts.SYSTEM_FAILED_ATTRIBUTE_PATH] = Boolean(False)
```
Comment (aniezurawski) [inline neptune/new/internal/init_impl.py:218]: Just False should work as well.

File: neptune/new/internal/utils/traceback_job.py (hunk 2 above)
Comment (aniezurawski) [reply neptune/new/internal/utils/traceback_job.py]: Just True. It should not be necessary to pack it.

File: neptune/new/attributes/constants.py [hunk 6]
```diff
@@ -21,6 +21,7 @@
 MONITORING_ATTRIBUTE_SPACE = 'monitoring/'
 MONITORING_STDERR_ATTRIBUTE_PATH = f'{MONITORING_ATTRIBUTE_SPACE}stderr'
 MONITORING_STDOUT_ATTRIBUTE_PATH = f'{MONITORING_ATTRIBUTE_SPACE}stdout'
+MONITORIC_TRACEBACK = f'{MONITORING_ATTRIBUTE_SPACE}traceback'
```
Comment (aniezurawski) [inline neptune/new/attributes/constants.py:24]: MONITOR**ING**_TRACEBACK_ATTRIBUTE_PATH

File: neptune/new/internal/utils/traceback_job.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        self._uuid = uuid.uuid4()
+        self._started = False
+        self._path = path
+
+    def start(self, run: 'Run'):
+        if not self._started:
+            path = self._path
+
+            def log_traceback(stacktrace_lines: List[str]):
+                run[path].log(stacktrace_lines)
```
[the remark this answers is not in the document]
Comment (aniezurawski) [reply neptune/new/internal/utils/traceback_job.py:45]: What about this comment?

File: neptune/new/internal/utils/uncaught_exception_handler.py [hunk 8]
```diff
@@ -31,31 +32,36 @@ class UncaughtExceptionHandler:
     def __init__(self):
         self._previous_uncaught_exception_handler = None
         self._handlers = dict()
+        self.__lock__ = threading.Lock()
```
Comment (aniezurawski) [inline neptune/new/internal/utils/uncaught_exception_handler.py:35]: This naming convention is reserved for built-in things.

**Review by aniezurawski — CHANGES_REQUESTED** [review]

File: neptune/new/internal/utils/traceback_job.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (HubertJaworski) [reply neptune/new/internal/utils/traceback_job.py:45]: I chose not to do this at this point. We don't know what is valuable for user and stuff like hostname should be logged anyway in other attributes. The name of series is literally `traceback`, and for me, adding random stuff there will make this series harder to read

[the remark this answers is not in the document]
Comment (aniezurawski) [reply neptune/new/internal/utils/traceback_job.py:45]: I'm not sure if this is something to discuss. It make sense to have such information.
@Herudaio ?

**Review by aniezurawski — APPROVED** [review]

**Pull request merged**